Create poem from submission form

whyderrick opened the pull request:
This PR is concerned with creating a new poem from the form to create a new submission.	The approach I've taken here works, but I'm skeptical of how sound it is. Essentially, I manually create the associated poem in `SubmissionsController` instead of using `accepts_nested_attributes_for` or creating a separate form object. I tried both of those approaches and failed with them.
 - I wasn't able to get `accepts_nested_attributes_for` wasn't properly assigning the `current_user` to the newly created poem. This approach allowed me to modify the params directly. I might've been able to use a `before_save` callback, but I recall thoughtbot having a style premise of avoiding them when possible. 
- I started to extract this to a service object, but found that the logic there was no different from what's currently in the controller and that it didn't fit with the eventual roadmap item of adding multiple new poems over AJAX while creating a new submission. 
-Eventually, I'd like to create this as a transaction or otherwise ensure that I don't have poems without a submission that are created if the submission fails to create. I'm sure there's a pattern I'm not using for this type of creation. The urge towards a transaction might be manageable by only creating new poems over AJAX.

whyderrick committed:
Update code to utilize namespacing with statuses

This the last PR changed the names on some model constants but didn't
update those references throughout the code. This updates that
oversight.

whyderrick committed:
Create new poem from submission form

This commit allows a user to create a new poem when creating a
submission. There were a few failed approaches before this commit.
 - I wasn't able to get `accepts_nested_attributes_for` to properly assign
the `current_user` to the newly created poem and strong params didn't
provide an easy way to access the nested attributes for the new poem.
- Extracting this to a service object didn't address that problem and
didn't seem to fit with the eventual roadmap item of adding poems
through AJAX.
- Eventually, I'd like to create this as a transaction, but there's a
pattern I'm going to want to use of a misconception I'll want to
correct, because in thinking about the ability to add multiple poems via
AJAX, I was thinking towards a switch in the controller based on what
the request contains. That said, an AJAX option that just builds and
returns poems on the submission will work the way I want.

gylaz commented on app/controllers/submissions_controller.rb:
```diff
@@ -4,7 +4,8 @@ def new
   def create
-    @submission = Submission.create(submission_params_with_user)
+    @submission = current_user.submissions.create(submission_params)
+    @submission.persisted? && @submission.poems.create(extract_poem_params)
```
You could probably rename `extract_poem_params` to `poem_params`.

gylaz commented on app/controllers/submissions_controller.rb:
```diff
@@ -24,7 +26,8 @@ def submission_params
-    submission_params.merge(user_id: current_user.id)
+  def extract_poem_params
+    @_poem_params ||= params[:submission].delete(:poem)
+    @_poem_params.permit(:title, :status).merge(user_id: current_user.id)
```
Any reason, this wouldn't work?
```rb
params.require(:submission).require(:poem).
  permit(:title, :status).
  merge(user_id: current_user.id)
```

gylaz commented on app/controllers/submissions_controller.rb:
```diff
@@ -16,6 +17,7 @@ def show
   private
 
   def submission_params
+    extract_poem_params
```
Any reason to call this method here?

gylaz reviewed (commented):
Typically, `accepts_nested_attributes` is what is used in this situation.

whyderrick committed:
`SubmissionForm` uses `accepts_nested_attributes_for`

This commit updates the `SubmissionsController` to use
`accepts_nested_attributes_for` instead of the previous workaround for
it.

Lessons
====
- Prior to this commit, I hadn't declared the `has_many_through`
relationship for `Submission` and `Poem`. A `has_many_through` that uses
`accepts_nested_attributes_for` two nested attributes statements: One on
the model that's accepting the other attributes and **one on the join
model**.
- Rails doesn't automatically assign the same `user_id` to all nested
items with an `accepts_nested_attributes_for`. The
`SubmissionController` here takes an extra step to ensure all poems in
the submission belong to `current_user`.
- `reject_if` is the method to use if every submission with
`accepts_nested_attributes_for` if the nested model is optional.

jcmorrow commented on app/controllers/submissions_controller.rb:
```diff
@@ -1,11 +1,16 @@
 class SubmissionsController < ApplicationController
   def new
     @submission = Submission.new
+    @submission.submission_packets.build.build_poem
```
This _technically_ violates demeter. It's not necessarily a problem, but it might be worth wrapping in a method on `Submission`.

jcmorrow commented on app/controllers/submissions_controller.rb:
```diff
@@ -1,11 +1,16 @@
-    @submission = Submission.create(submission_params_with_user)
+    @submission = current_user.submissions.build(submission_params)
+    @submission.submission_packets.map do |packet|
+      packet.poem.user_id = current_user.id
```
I could be wrong here, but it seems like if you have build the `submission` from `current_user` you shouldn't also need to assign it here.

whyderrick replied:
I thought so too, but I'm getting the database complaining that I'm trying to create a record in  `poems` with no `user_id` value.

Possible I'm missing something somewhat obvious. If you've got a few minutes to pair and spotcheck, that'd be great.

jcmorrow commented on app/models/submission.rb:
```diff
@@ -15,7 +15,28 @@ class Submission < ApplicationRecord
+      elsif value.blank?
+        true
+      else
+        false
```
this is one place that you should probably return early. Might be something helpful to TDD together.

jcmorrow commented on app/models/user.rb:
```diff
@@ -2,4 +2,5 @@ class User < ApplicationRecord
   include Clearance::User
 
   has_many :poems, dependent: :destroy
```
`through: :submissions`?

whyderrick replied:
I don't think so here. The user could feasibly create a poem and track it here before it has any associated submissions.

whyderrick replied:
For documentation: This really points to the fact that the current logic creates duplication around the user. While a poem might exist without a submission, it might make sense to disallow a submission with no poems, in which case I can make the submission's user come from it's poems (any one of them at this point).

jcmorrow reviewed (commented):
Seems like maybe the `empty_packet?` stuff needs some work. Other than that it looks 😎

gylaz commented on app/controllers/submissions_controller.rb:
```diff
@@ -1,11 +1,16 @@
   def create
-    @submission = Submission.create(submission_params_with_user)
+    @submission = current_user.submissions.build(submission_params)
+    @submission.submission_packets.map do |packet|
```
`map` is typically reserved for situations where you want to capture the result, if you don't intend to do that then `each` communicates that better.

whyderrick replied:
:+1: noted. Confused "making a change to the object" for "capturing the result".

gylaz commented on spec/features/submission_spec.rb:
```diff
@@ -23,14 +23,46 @@
+
+      expect(page).to have_text(submission_title)
+      expect(page).to have_text(new_poem_title)
+
```
✂️ ?

gylaz commented on app/models/submission.rb:
```diff
@@ -15,7 +15,28 @@ class Submission < ApplicationRecord
+    attributes[:poem_attributes].each_value.all?(&:blank?)
+  end
+
+  def deep_blank?(attributes)
```
As @jcmorrow pointed out, I would love to see some tests of this method. As is, I believe the method isn't going to do what you want it to do -- it will return `attributes`.
```rb
> {a: 1, b: { c: 3, d: 4}}.each_pair { |k, v| true }

{:a=>1, :b=>{:c=>3, :d=>4}}
```
